components/nimbus/src/behavior.rs

```diff
@@ -0,0 +1,415 @@
+    }
+}
+
+impl IntervalConfig {
+    fn new(bucket_count: usize, interval: Interval) -> Self {
+        Self {
+            bucket_count,
+            interval,
+        }
+    }
+}
+
+pub struct IntervalData {
+    buckets: VecDeque<u64>,
+    bucket_count: usize,
+    starting_instant: DateTime<Utc>,
+}
+
+impl Default for IntervalData {
+    fn default() -> Self {
+        Self::new(1)
+    }
+}
+
+impl IntervalData {
+    fn new(bucket_count: usize) -> Self {
+        let mut data = Self {
+            buckets: VecDeque::new(),
+            bucket_count,
+            starting_instant: Utc::now(),
+        };
+        data.buckets.push_front(0);
+        data
+    }
+
+    pub fn from(
+        buckets: VecDeque<u64>,
+        bucket_count: usize,
+        starting_instant: DateTime<Utc>,
+    ) -> Self {
+        Self {
+            buckets,
+            bucket_count,
+            starting_instant,
+        }
+    }
+
+    fn increment(&mut self) -> Result<()> {
+        match self.buckets.front_mut() {
+            Some(x) => *x += 1,
+            None => {
+                return Err(NimbusError::BehaviorError(BehaviorError::InvalidState(
+                    "Interval buckets cannot be empty".to_string(),
+                )))
+            }
+        };
+        Ok(())
+    }
+
+    fn rotate(&mut self, num_rotations: u64) -> Result<()> {
+        for _ in 1..=num_rotations {
+            self.buckets.push_front(0)
+        }
+        if self.buckets.len() > self.bucket_count {
+            self.buckets.drain(self.bucket_count..);
+        }
+        Ok(())
+    }
+}
+
+fn date_diff(a: DateTime<Utc>, b: DateTime<Utc>) -> u32 {
+    (a.date() - b.date()).num_days().try_into().unwrap()
+}
+
+struct SingleIntervalCounter {
+    data: IntervalData,
+    config: IntervalConfig,
+}
+
+impl SingleIntervalCounter {
+    pub fn new(config: IntervalConfig) -> Self {
+        Self {
+            data: IntervalData::new(config.bucket_count),
+            config,
+        }
+    }
+
+    pub fn from(data: IntervalData, config: IntervalConfig) -> Self {
+        Self { data, config }
+    }
+
+    pub fn increment(&mut self) -> Result<()> {
+        self.data.increment()
+    }
+
+    pub fn maybe_advance(&mut self, now: DateTime<Utc>) -> Result<()> {
+        let then = self.data.starting_instant;
+        let rotations: u32 = match self.config.interval {
+            Interval::Minutes => now.minute() - then.minute(),
+            Interval::Hours => now.hour() - then.hour(),
+            Interval::Days => date_diff(now, then),
+            Interval::Weeks => date_diff(now, then) / 7,
+            Interval::Months => date_diff(now, then) / 28,
+            Interval::Years => date_diff(now, then) / 365,
+        };
+        if rotations > 0 {
+            return self.data.rotate(rotations.into());
+        }
+        Ok(())
+    }
+}
+
+struct MultiIntervalCounter {
+    intervals: HashMap<IntervalName, SingleIntervalCounter>,
```

Comment:
See my comment about `Interval` and `IntervalName` above, but after looking at this I don't think that we are really using this `IntervalName` inside of `MultiIntervalCounter` so maybe the key for our HashMap could just be the `Interval` enum instead of having to have a separate "named" one that should just be corresponding to the `Interval` anyway. Does this make sense?